### ISSUE TITLE
Spikes of TCP connections before getting stuck

### ISSUE THREAD

**Issue by tstdin** [input-output-hk/jormungandr#1521]
**Describe the bug**
I am seeing rapid increase of TCP connection right before the pool gets stuck (plus 100-200 connections).
![image](https://user-images.githubusercontent.com/25065143/71955812-a0045180-31e9-11ea-9c69-5a6d6462bbf4.png)

Interesting is, this happens periodically every ~ 30 minutes. The drop of connections to the bottom is the automatic restart of the pool (I have a script that detects if the pool is stuck and then restarts it).

I am not sure if this is a Jormungandr problem or rather some malicious behaviour by 3rd party (DoS attack).

**Mandatory Information**
I have observed this behaviour in v0.8.5 and also in version built from master 35a60a3c.

**To Reproduce**
1. Start Jormungandr.
2. Wait slightly over 30 minutes.
3. Pool gets restarted.
4. Repeat from 1.

**Expected behavior**
Run Jormungandr without getting stuck.

**Additional context**
N/A

**Comment by manveru** [comment]
The underlying cause is the same as https://github.com/input-output-hk/jormungandr/issues/1459

**Comment by tstdin** [comment]
> The underlying cause is the same as #1459

In that case the issue #1459 is still not fixed, as I was experiencing that behaviour also on 35a60a3.

If it would help, I can provide a packet capture of one of those incidents. But I will share it only privately, as it is not anonymized.

**Comment by mzabaluev** [comment]
This might have the same cause as #1459.

**Comment by tstdin** [comment]
With #1523 applied, it behaves still the same. The last three runs on the following graph are with those changes.
![image](https://user-images.githubusercontent.com/25065143/72012096-8d2a6500-325b-11ea-9c15-8cc6e580a404.png)

**Comment by jalora8** [comment]
I am experiencing the same problem.

**Required Info**
```
jormungandr --full-version
jormungandr 0.8.5 (HEAD-35a60a3c, release, linux [x86_64]) - [rustc 1.40.0 (73528e339 2019-12-16)]
```

`Max Connections set to: 1024
`
```
cat /proc/version
Linux version 4.15.0-1057-aws (buildd@lgw01-amd64-030) (gcc version 7.4.0 (Ubuntu 7.4.0-1ubuntu1~18.04.1)) #59-Ubuntu SMP Wed Dec 4 10:02:00 UTC 2019
```

```
Machine info:
Intel(R) Xeon(R) CPU E5-2676 v3 @ 2.40GHz (2 cores)
RAM: 4 GB
```

**Observations:** Increase in CPU, culminating with spikes in cpu usage and associated spikes in tcp connections, socket memory, and tcp memory. Jormungandr terminates in error via a "network" panic. See below.

```
Jan 09 19:32:53.982 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 09 19:32:54.152 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 09 19:32:54.187 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
thread 'network1' panicked at 'called `Result::unwrap()` on an `Err` value: SpawnError { is_shutdown: true }', src/libcore/result.rs:1165:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
Jan 09 20:56:17.226 CRIT Task panicked, task: network
```

<img width="601" alt="CPU spike" src="https://user-images.githubusercontent.com/58919321/72106654-5d2fa000-32f5-11ea-9da9-e4509d61bdbe.png">

<img width="620" alt="TCP spikes" src="https://user-images.githubusercontent.com/58919321/72106581-3e310e00-32f5-11ea-9ebb-eec79bd95b54.png">

<img width="624" alt="TCP mem spike" src="https://user-images.githubusercontent.com/58919321/72106593-4426ef00-32f5-11ea-8c32-96ea45c19031.png">

<img width="611" alt="socket mem spike" src="https://user-images.githubusercontent.com/58919321/72106667-6456ae00-32f5-11ea-845d-da1ad05a8011.png">

**Comment by jalora8** [comment]
**Further update:**
After adjusting my config.yaml to remove `listen_address` and `public_address`, I am no longer observing the spikes and my cpu is behaving well. Also, I have set my `max_connections=512`. Node is stable.

**Note that the far right corner (after 17:15) is the behavior that I am seeing.**

<img width="609" alt="cpu basic" src="https://user-images.githubusercontent.com/58919321/72116880-3e3f0700-3311-11ea-9d62-7ed883429d2b.png">

<img width="602" alt="tcp connect" src="https://user-images.githubusercontent.com/58919321/72116891-4434e800-3311-11ea-93b0-bf1f09be2e1e.png">

<img width="618" alt="Socket TCP" src="https://user-images.githubusercontent.com/58919321/72116902-4bf48c80-3311-11ea-930e-3ef7fb08cc20.png">

<img width="615" alt="socket mem" src="https://user-images.githubusercontent.com/58919321/72116906-5020aa00-3311-11ea-92ad-88bb6b61c723.png">

**Comment by hemants1** [comment]
But isn't listen_address be configured to receive block announcements being published to all the nodes in the network to keep the blockchain upto date? but otherwise how would keep your node in sycn other than your node connecting to other node pull the last update on blocks/slots.   What happens if every node in the network has this listen_address removed? will any node be able to keep up - updated & synched????

**Comment by hemants1** [comment]
> **Further update:**
> After adjusting my config.yaml to remove `listen_address` and `public_address`, I am no longer observing the spikes and my cpu is behaving well. Also, I have set my `max_connections=512`. Node is stable.
> 
> **Note that the far right corner (after 17:15) is the behavior that I am seeing.**
> 
> <img alt="cpu basic" width="609" src="https://user-images.githubusercontent.com/58919321/72116880-3e3f0700-3311-11ea-9d62-7ed883429d2b.png">
> 
> <img alt="tcp connect" width="602" src="https://user-images.githubusercontent.com/58919321/72116891-4434e800-3311-11ea-93b0-bf1f09be2e1e.png">
> 
> <img alt="Socket TCP" width="618" src="https://user-images.githubusercontent.com/58919321/72116902-4bf48c80-3311-11ea-930e-3ef7fb08cc20.png">
> 
> <img alt="socket mem" width="615" src="https://user-images.githubusercontent.com/58919321/72116906-5020aa00-3311-11ea-92ad-88bb6b61c723.png">

How are you able to keep your node in sync? by just connecting to peers in network? Are you not receiving any "failed to connect to peer, reason: Connection refused (os error 111), " if the peer has removed the "LISTEN_ADDRESS" settings?

**Comment by hemants1** [comment]
> `Max Connections set to: 1024 `

Also as per my understanding, Max Connections is for inbound connections to your node...! I believe if you remove LISTEN_ADDRESS then this parameter is not applicable..! and obviously your node will not be bombarded with inbound connections from other node in the network hence CPU and MEM remain utilized very low...!

**Comment by mzabaluev** [comment]
> After adjusting my `config.yaml` to remove `listen_address` and `public_address`, I am no longer observing the spikes and my cpu is behaving well.

This suggests that the problem in the gRPC server implementation; one fix is pending in #1531.

**Comment by jalora8** [comment]
I'm still seeing this on `b45d2355`. No longer observed the original panic (although I may not have run it long enough to get there). I set `max_connections=512`, `max_unreachable_nodes_to_connect_per_event=2`, and `gossip_interval=3s` . See below snapshots for more details. Hope it helps.

**Gradual build up in CPU load but no longer seeing a spike w/ associated panic**
<img width="606" alt="CPU" src="https://user-images.githubusercontent.com/58919321/72201239-85172480-3417-11ea-9ccc-91f26a2e0f79.png">

**Same spike on socket TCP and TCP connections**
<img width="618" alt="Sockstat TCP" src="https://user-images.githubusercontent.com/58919321/72201264-b0017880-3417-11ea-8ba7-70f41c7e8499.png">
<img width="613" alt="TCP spike" src="https://user-images.githubusercontent.com/58919321/72201339-711ff280-3418-11ea-88a2-8cbe05716027.png">


**AND... Socket memory**
<img width="617" alt="Socket Memory" src="https://user-images.githubusercontent.com/58919321/72201271-ca3b5680-3417-11ea-955b-23f51168a122.png">

**During TCP and socket memory spike, scraper was having difficulty getting block height and node becomes stuck**
<img width="620" alt="blockheight" src="https://user-images.githubusercontent.com/58919321/72201304-feaf1280-3417-11ea-89df-7e3d48f5f140.png">

**Comment by jalora8** [comment]
Looking further into my logs I saw the following

```
Jan 11 08:07:50 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:07:50.087 CRIT Task panicked, task: network
Jan 11 08:07:50 ip-172-31-2-14 startFRG.sh[24483]: thread 'network1' panicked at 'called `Result::unwrap()` on an `Err` value: SpawnError { is_shutdown: true }', src/libcore/result.rs:1165:5
Jan 11 08:07:48 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:07:48.778 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:07:48 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:07:48.139 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:07:45 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:07:45.546 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:07:43 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:07:43.248 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:07:42 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:07:42.447 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:07:42 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:07:42.447 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:07:42 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:07:42.446 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:07:42 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:07:42.445 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
.........
Jan 11 08:05:44 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:05:44.067 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:05:44 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:05:44.066 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:05:44 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:05:44.066 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:05:44 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:05:44.065 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:05:44 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:05:44.065 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:05:44 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:05:44.064 WARN response canceled by the processing task, request: PushHeaders, sub_task: server, task: network
Jan 11 08:05:17 ip-172-31-2-14 startFRG.sh[24483]: Jan 11 08:05:17.114 CRIT Task panicked, task: block
Jan 11 08:05:17 ip-172-31-2-14 startFRG.sh[24483]: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
Jan 11 08:05:17 ip-172-31-2-14 startFRG.sh[24483]: thread 'block0' panicked at 'referential integrity failure in CandidateForest', src/libcore/option.rs:1185:5
```

TLDR; It looks like I got a block panic prior to the network panic.

**Comment by hemants1** [comment]
How long your are able to keep the node up and running with these settings before it is stuck?


Regards,
Hemant Srinivas
1 262-290-6475 (M)
1 414-524-5536 (O)

________________________________

**Comment by jalora8** [comment]
> How long your are able to keep the node up and running with these settings before it is stuck?

Same as what I experienced in my first comment... a period of about 1 hour.

**Comment by skokasik** [comment]
I notice the same behaviour. with settings below
![screen](https://user-images.githubusercontent.com/2031606/72264109-19ba9600-3655-11ea-88d2-3ec600b4b3c4.png)


  public_address: "/ip4/..../tcp/3000"
  listen_address: "/ip4/0.0.0.0/tcp/3000"
  public_id: ....
  max_unreachable_nodes_to_connect_per_event: 1
  max_connections: 256
rest:
  listen: 127.0.0.1:3100
leadership:
  log_ttl: 24h
  garbage_collection_interval: 10m
storage:...
explorer:
  enabled: false
mempool:
  fragment_ttl: 30m
  log_ttl: 24h
  garbage_collection_interval: 10m
  pool_max_entries: 100000
  log_max_entries: 1000000

**Comment by manveru** [comment]
Same issue with 0.8.6, just that it only took about 5 minutes to reach the impasse.

**Comment by jalora8** [comment]
CPU load is still ramping up. TCP connections and TCP socket memory still spiking as well on 0.8.6.

<img width="527" alt="Screen Shot 2020-01-15 at 1 15 57 PM" src="https://user-images.githubusercontent.com/58919321/72463632-4c6da700-3799-11ea-8c8f-0d011f7295c2.png">

<img width="534" alt="Screen Shot 2020-01-15 at 1 16 21 PM" src="https://user-images.githubusercontent.com/58919321/72463640-4f689780-3799-11ea-926e-ece555d89871.png">

**Comment by mmahut** [comment]
I can confirm we are seeing this on 0.8.6. as well.

**Comment by cardanians** [comment]
confirming too!

**Comment by hemants1** [comment]
Confirming too. My Node settings are:

  topics_of_interest:
    blocks: high
    messages: high
  max_connections: 256
  max_unreachable_nodes_to_connect_per_event: 2
  gossip_interval: 1s
  policy:
    quarantine_duration: 30m

mempool:
  fragment_ttl: 30m
  log_max_entries: 300000
  pool_max_entries: 30000
  log_ttl: 1h
  garbage_collection_interval: 10m
leadership:
  log_ttl: 1h
  garbage_collection_interval: 10m

**Comment by Linicks** [comment]
I can also confirm that I'm seeing this on 8.6 release on Ubuntu.
![Screenshot from 2020-01-15 17-20-28](https://user-images.githubusercontent.com/295692/72482616-95ceee00-37bb-11ea-806c-d614213c041c.png)

**Comment by kakerley** [comment]
I am also seeing this on 8.6 and all previous versions. Running Ubuntu on a RockPi. Coincides with the following netstats and falling behind in block height. Can't keep this thing going for longer than 10 minutes it seems.

root@rockpi:~/jormungandr-node# netstat -tn | tail -n +3 | awk "{ print \$6 }" | sort | uniq -c | sort -n
      1 LAST_ACK
     20 TIME_WAIT
     66 SYN_SENT
    173 CLOSE_WAIT
   1758 ESTABLISHED
root@rockpi:~/jormungandr-node# netstat -tn | tail -n +3 | awk "{ print \$6 }" | sort | uniq -c | sort -n
      1 LAST_ACK
     12 TIME_WAIT
     60 SYN_SENT
    227 CLOSE_WAIT
   1689 ESTABLISHED
root@rockpi:~/jormungandr-node# netstat -tn | tail -n +3 | awk "{ print \$6 }" | sort | uniq -c | sort -n
     13 TIME_WAIT
     58 SYN_SENT
    227 CLOSE_WAIT
   1690 ESTABLISHED

**_systemctl service stops jormungandr here:_**

root@rockpi:~/jormungandr-node# netstat -tn | tail -n +3 | awk "{ print \$6 }" | sort | uniq -c | sort -n
      1 CLOSING
      1 LAST_ACK
     12 ESTABLISHED
    121 FIN_WAIT1
    564 TIME_WAIT
root@rockpi:~/jormungandr-node# netstat -tn | tail -n +3 | awk "{ print \$6 }" | sort | uniq -c | sort -n
      3 TIME_WAIT
     12 ESTABLISHED

**_Here is the block height trail:_** 

No block height
No block height
No block height
2020-01-15 23:45:18 New block height: "100729"
2020-01-15 23:45:38 Current block height: "100729" - No new block for 20 sec.
2020-01-15 23:45:58 New block height: "100840"
2020-01-15 23:46:18 New block height: "100872"
2020-01-15 23:46:38 New block height: "100938"
2020-01-15 23:46:58 New block height: "101098"
2020-01-15 23:47:19 New block height: "101130"
2020-01-15 23:47:39 New block height: "101194"
2020-01-15 23:47:59 New block height: "101234"
2020-01-15 23:48:19 New block height: "101330"
2020-01-15 23:48:39 New block height: "101362"
2020-01-15 23:48:59 New block height: "101507"
2020-01-15 23:49:19 New block height: "101577"
2020-01-15 23:49:39 New block height: "101673"
2020-01-15 23:49:59 New block height: "101769"
2020-01-15 23:50:19 New block height: "101801"
2020-01-15 23:50:39 New block height: "101888"
2020-01-15 23:51:00 Current block height: "101888" - No new block for 20 sec.
2020-01-15 23:51:20 Current block height: "101888" - No new block for 40 sec.
2020-01-15 23:51:40 Current block height: "101888" - No new block for 60 sec.
2020-01-15 23:52:00 Current block height: "101888" - No new block for 80 sec.
2020-01-15 23:52:20 Current block height: "101888" - No new block for 100 sec.
2020-01-15 23:52:40 Current block height: "101888" - No new block for 120 sec.
2020-01-15 23:53:00 Current block height: "101888" - No new block for 140 sec.
2020-01-15 23:53:20 Current block height: "101888" - No new block for 160 sec.
2020-01-15 23:53:40 Current block height: "101888" - No new block for 180 sec.
2020-01-15 23:54:00 New block height: "101889"
2020-01-15 23:54:20 New block height: "101890"
2020-01-15 23:54:41 Current block height: "101890" - No new block for 20 sec.
2020-01-15 23:55:01 Current block height: "101890" - No new block for 40 sec.
2020-01-15 23:55:21 Current block height: "101890" - No new block for 60 sec.
2020-01-15 23:55:41 Current block height: "101890" - No new block for 80 sec.
2020-01-15 23:56:01 Current block height: "101890" - No new block for 100 sec.
2020-01-15 23:56:21 Current block height: "101890" - No new block for 120 sec.
2020-01-15 23:56:41 Current block height: "101890" - No new block for 140 sec.
2020-01-15 23:57:01 Current block height: "101890" - No new block for 160 sec.
2020-01-15 23:57:21 Current block height: "101890" - No new block for 180 sec.
2020-01-15 23:57:41 Current block height: "101890" - No new block for 201 sec.
2020-01-15 23:58:02 Current block height: "101890" - No new block for 221 sec.
2020-01-15 23:58:22 Current block height: "101890" - No new block for 241 sec.
2020-01-15 23:58:42 Current block height: "101890" - No new block for 261 sec.
2020-01-15 23:59:02 Current block height: "101890" - No new block for 281 sec.
2020-01-15 23:59:22 Current block height: "101890" - No new block for 301 sec.
2020-01-15 23:59:42 Current block height: "101890" - No new block for 321 sec.
2020-01-16 00:00:02 Current block height: "101890" - No new block for 341 sec.
2020-01-16 00:00:22 Current block height: "101890" - No new block for 361 sec.

2020-01-16 00:00:22 Stopping jormungandr
Success
Sleeping for 5 sec.
Restarting shelley.service
Sleeping for 60 sec.
No block height
No block height

**Comment by cardanians** [comment]
confirming - 0.8.6.
![obrazek](https://user-images.githubusercontent.com/58650764/72527194-7f32a080-3868-11ea-9e00-f823cffdab2b.png)

**Comment by mzabaluev** [comment]
One helpful way to debug this is to obtain a backtrace for all threads when the CPU busy-looping is reproduced, by performing either of these:

* Attach to the jormungandr process with the debugger and dump the backtraces for all threads;
* On Linux or Mac OS, kill jormungandr with `kill -ABRT` and dump the backtraces with the debugger from the core dump (which needs to be enabled).

**Comment by mzabaluev** [comment]
Also, dump the sockets that the jormungandr process has open at this time, with `lsof -nP` or `ss`. Please make sure to filter the output so that only TCP sockets used by jormungandr are included.

**Comment by stanfieldr** [comment]
[logs.zip](https://github.com/input-output-hk/jormungandr/files/4072286/logs.zip)

**Comment by slackwaree** [comment]
public_id:
max_unreachable_nodes_to_connect_per_event: 10
max_connections: 480

Are all invalid options both in p2p and in the global section so how did you actually put these in? Maybe use the insert code
`<?php phpinfo(); ?>`

next time when you post, it works as intended unlike your rant.
